packages/react-server/src/ReactFizzServer.js

```diff
@@ -97,119 +212,580 @@ function renderNode(request: Request, node: ReactNodeList): void {
   if (typeof type === 'function') {
     try {
       const result = type(props);
-      renderNode(result);
+      renderNode(request, parentBoundary, segment, result);
     } catch (x) {
       if (typeof x === 'object' && x !== null && typeof x.then === 'function') {
         // Something suspended, we'll need to create a new segment and resolve it later.
-        request.pendingChunks++;
-        const newSegment = createSegment(request, node);
-        const ping = newSegment.ping;
+        const insertionIndex = segment.chunks.length;
+        const newSegment = createPendingSegment(request, insertionIndex, null);
+        const suspendedWork = createSuspendedWork(
+          request,
+          node,
+          parentBoundary,
+          newSegment,
+          null,
+        );
+        const ping = suspendedWork.ping;
         x.then(ping, ping);
         // TODO: Emit place holder
       } else {
-        reportError(request, x);
-        // TODO: Emit client render signal
+        // We can rethrow to terminate the rest of this tree.
+        throw x;
       }
     }
   } else if (typeof type === 'string') {
-    request.completedChunks.push(formatChunk(type, props));
+    // TODO: Split this to start/end and process children.
+    segment.chunks.push(formatChunk(type, props));
+  } else if (type === REACT_SUSPENSE_TYPE) {
+    // Each time we enter a suspense boundary, we split out into a new segment for
+    // the fallback so that we can later replace that segment with the content.
```

Comment:
What happens if a fallback itself Suspends? Did you have to build anything special for this case?

Reply:
It should just work if my calculations are correct. (Ofc, I have tested any of this.)

A fallback is rendered just like plain children in the parent of the Suspense boundary would be. So they end up being children of the fallback.

You can also have a suspense boundary within the fallback. If the main content gets deleted before that resolves, then it won't find the fallback in getElementById so it won't insert it. This is no different than if React deleted the tree such as if the user switches tab before the full page has loaded.

However, there is an outstanding to do here to cancel the fallback if the main content succeeds. Right now we have to compute the fallback before showing the main content which is not ideal. So I need to build in cancelation. That doesn't affect the suspended fallback though.

The fallback suspending adds to the ref count (pendingWork) of the parent so therefore we end up waiting on the fallback even if the main content succeeds. I explored other options but I think cancelation is the right strategy to solve this.